src/main/java/org/openrewrite/java/migrate/AddMissingMethodImplementation.java

```diff
@@ -0,0 +1,151 @@
+
+    // All recipes must be serializable. This is verified by RewriteTest.rewriteRun() in your tests.
+    @JsonCreator
+    public AddMissingMethodImplementation(@NonNull @JsonProperty("fullyQualifiedClassName") String fullyQualifiedClassName,
+                                          @NonNull @JsonProperty("methodPattern") String methodPattern,
+                                          @NonNull @JsonProperty("methodTemplateString") String methodTemplateString) {
+        this.fullyQualifiedClassName = fullyQualifiedClassName;
+        this.methodPattern = methodPattern;
+        this.methodTemplateString = methodTemplateString;
+    }
+
+    @Override
+    public String getDisplayName() {
+        return "Adds missing method implementations.";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Check for missing methods required by interfaces and adds them.";
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return Preconditions.check(new UsesType<>(fullyQualifiedClassName, true), new ClassImplementationVisitor());
+    }
+
+    public class ClassImplementationVisitor extends JavaIsoVisitor<ExecutionContext> {
+        private final JavaTemplate methodTemplate = JavaTemplate.builder(methodTemplateString).build();
+        private final MethodMatcher methodMatcher = new MethodMatcher(methodPattern, true);
+
+        public boolean matchesInterface(JavaType.Class type) {
+            if (type != null) {
+                if (type.getFullyQualifiedName().equals(fullyQualifiedClassName)) {
+                    return true;
+                }
+
+                // check for matches on super interface
+                List<JavaType.FullyQualified> superInterfaces = type.getInterfaces();
+                boolean foundOnSuperInterface = false;
+                for (JavaType.FullyQualified superInterface : superInterfaces) {
+                    if (matchesInterface((JavaType.Class) superInterface)) {
+                        foundOnSuperInterface = true;
+                        break;
+                    }
+                }
+                return foundOnSuperInterface;
+            }
+
+            return false;
+        }
+
+        public boolean implementsInterface(J.ClassDeclaration classDecl) {
+            List<TypeTree> implementedClasses = classDecl.getImplements();
+            if (implementedClasses != null) {
+                for (TypeTree implementedClass : implementedClasses) {
+                    JavaType.Class type = (JavaType.Class) implementedClass.getType();
+                    return matchesInterface(type);
+                }
+            }
+
+            // also check extended class to handle abstract classes
+            TypeTree extendedClass = classDecl.getExtends();
+            if (extendedClass != null) {
+                JavaType.Class type = (JavaType.Class) extendedClass.getType();
+                return matchesInterface(type);
+            }
```

Comment:
We likely also don't want to add methods to classes that themselves are abstract, instead pushing that down into actual implementations. Or what are your thoughts on that?

Reply:
As a continuation of https://github.com/openrewrite/rewrite-migrate-java/pull/312/files#r1364317262